specs/lightclient/beacon-chain.md

```diff
@@ -0,0 +1,201 @@
+
+#### `BeaconBlock`
+
+```python
+class BeaconBlock(phase0.BeaconBlock):
+    # Sync committee aggregate signature
+    sync_committee_bits: Bitvector[SYNC_COMMITTEE_SIZE]
+    sync_committee_signature: BLSSignature
+```
+
+#### `BeaconBlockHeader`
+
+```python
+class BeaconBlockHeader(phase0.BeaconBlockHeader):
+    # Sync committee aggregate signature
+    sync_committee_bits: Bitvector[SYNC_COMMITTEE_SIZE]
+    sync_committee_signature: BLSSignature
+```
+
+
+#### `BeaconState`
+
+```python
+class BeaconState(phase0.BeaconState):
+    # Sync committees
+    current_sync_committee: SyncCommittee
+    next_sync_committee: SyncCommittee
+```
+
+### New containers
+
+#### `SyncCommittee`
+
+```python
+class SyncCommittee(Container):
+    pubkeys: Vector[BLSPubkey, SYNC_COMMITTEE_SIZE]
+    pubkey_aggregates: Vector[BLSPubkey, SYNC_COMMITTEE_SIZE // SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE]
+```
+
+## Helper functions
+
+### Beacon state accessors
+
+#### `get_sync_committee_indices`
+
+```python
+def get_sync_committee_indices(state: BeaconState, epoch: Epoch) -> Sequence[ValidatorIndex]:
+    """
+    Return the sync committee indices for a given state and epoch.
+    """
+    MAX_RANDOM_BYTE = 2**8 - 1
+    base_epoch = Epoch((max(epoch // EPOCHS_PER_SYNC_COMMITTEE_PERIOD, 1) - 1) * EPOCHS_PER_SYNC_COMMITTEE_PERIOD)
+    active_validator_indices = get_active_validator_indices(state, base_epoch)
+    active_validator_count = uint64(len(active_validator_indices))
+    seed = get_seed(state, base_epoch, DOMAIN_SYNC_COMMITTEE)
+    i, sync_committee_indices = 0, []
+    while len(sync_committee_indices) < SYNC_COMMITTEE_SIZE:
+        shuffled_index = compute_shuffled_index(uint64(i % active_validator_count), active_validator_count, seed)
+        candidate_index = active_validator_indices[shuffled_index]
+        random_byte = hash(seed + uint_to_bytes(uint64(i // 32)))[i % 32]
+        effective_balance = state.validators[candidate_index].effective_balance
+        if effective_balance * MAX_RANDOM_BYTE >= MAX_EFFECTIVE_BALANCE * random_byte:
+            sync_committee_indices.append(candidate_index)
+        i += 1
+    return sync_committee_indices
+```
+
+#### `get_sync_committee`
+
+```python
+def get_sync_committee(state: BeaconState, epoch: Epoch) -> SyncCommittee:
+    """
+    Return the sync committee for a given state and epoch.
+    """
+    indices = get_sync_committee_indices(state, epoch)
+    validators = [state.validators[index] for index in indices]
+    pubkeys = [validator.pubkey for validator in validators]
+    aggregates = [
+        bls.AggregatePKs(pubkeys[i:i + SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE])
+        for i in range(0, len(pubkeys), SYNC_COMMITTEE_PUBKEY_AGGREGATES_SIZE)
+    ]
+    return SyncCommittee(pubkeys, aggregates)
+```
+
+### Block processing
+
+```python
+def process_block(state: BeaconState, block: BeaconBlock) -> None:
+    phase0.process_block(state, block)
```

Comment:
Note that by `phase0.process_block`, it uses the phase 0 package.

1. The constants and the function calls `process_block` would be *phase 0 version*. It may cause some side effects that we may easily overlook.
2. You are passing `(lightclient_patch.state, lightclient_patch.block)` to `process_block(state: phase0.BeaconState, block: phase0.BeaconBlock)`.

Saving 3 lines probably is not worth it. 😅